tests/scale-tests/results/FQM_Latency_and_Queue_Depth_Results.md

```diff
@@ -0,0 +1,163 @@
+# FQM Latency & Queue Depth Performance
+
+**Objective:** Measure FQM (Fault Quarantine Module) performance for cordoning nodes in response to fatal health events.
+
+---
+
+## Test Environment
+
+**Cluster:** 1503 nodes (3 system + 1500 customer aws-cpu-m7i.xlarge nodes)  
+**NVSentinel Version:** v0.4.0  
+**MongoDB:** 3 replicas, 6Gi memory per replica  
+**Test Date:** December 4, 2025
+
+### Test Approach
+
+**Phase 1: End-to-End Tests** measure the complete pipeline latency from signal to cordon. This captures real-world performance including all component interactions.
+
+**Phase 2: Microbenchmarks** isolate individual pipeline stages using Prometheus metrics and direct MongoDB benchmarks to identify bottlenecks.
+
+### Pipeline Measured
+
+```
+┌─────────────┐    ┌───────────────────┐    ┌─────────┐    ┌─────────────┐    ┌───────────┐
+│  SIGUSR1    │───▶│ Event Generator   │───▶│ Platform│───▶│  MongoDB    │───▶│    FQM    │───▶ Node Cordoned
+│  (T0)       │    │ (gRPC/UDS)        │    │Connector│    │  (insert)   │    │ (change   │
+└─────────────┘    └───────────────────┘    └─────────┘    └─────────────┘    │  stream)  │
+                                                                              └───────────┘
+     [A]                  [B]                   [C]              [D]              [E]
+```
+
+**End-to-End Latency** = T(cordon) - T(SIGUSR1) — measures [A] through [E]
+
+---
+
+## FQM Latency Results
+
+| Scale | Nodes | Time to Complete | Peak Queue | Avg Rate |
+|-------|-------|------------------|------------|----------|
+| 10% | 150 | **64.5s** | 99 | 3.33/sec |
+| 25% | 375 | **156.8s** | 258 | 3.41/sec |
+| 50% | 750 | **418.6s** | 575 | 2.31/sec |
+
+**Test Methodology:** All SIGUSR1 signals are sent as fast as possible using a 50-worker pool, then we poll every 5 seconds to count cordoned nodes until all are complete. "Time to Complete" measures from first signal sent until last node cordoned.
+
+**Test Parameters:** Lightweight mode (worker pool), no stagger, Circuit breaker DISABLED, Node drainer DISABLED
+
+---
+
+## Key Findings
+
+- ✅ **100% cordoning success rate** at all scales (10%, 25%, 50%)
+- ✅ **~3 nodes/sec** cordoning rate at 10-25% load
+- ✅ **~2.3 nodes/sec** at 50% load (750 nodes)
+- ✅ **Event handling latency** stays consistent: P50 ~0.37s, P90 ~0.48s
+- ✅ **Peak backlog scales linearly** with load: 104 → 256 → 534
+- ✅ **Bottleneck is FQM** — Platform Connector writes to MongoDB in ~5ms with zero queue backlog
```

Comment:
I wouldn't say bottleneck. It has a queue and does one a time, so latency is expected to a certain extent. Can we rephrase this slightly so that the queue and one at a time is reflected?

Reply:
fixed!